src/mainboard/starfive/visionfive1/bt0/src/main.rs

```diff
@@ -0,0 +1,156 @@
+#![feature(naked_functions, asm_sym, asm_const)]
+#![no_std]
+#![no_main]
+
+use core::{
+    arch::{asm, global_asm},
+    panic::PanicInfo,
+};
+use init::{clock_init, gmac_init, iopad_init, rstgen_init, uart_write};
+use riscv;
+
+mod init;
+#[macro_use]
+mod log;
+
+global_asm!(include_str!("../start.S"));
+
+const STACK_SIZE: usize = 4 * 1024; // 4KiB
+
+#[link_section = ".bss.uninit"]
+static mut BT0_STACK: [u8; STACK_SIZE] = [0; STACK_SIZE];
+
+/// Set up stack and jump to executable code.
+///
+/// # Safety
+///
+/// Naked function.
+#[naked]
+#[export_name = "start"]
+#[link_section = ".text.entry"]
+pub unsafe extern "C" fn start() -> ! {
+    asm!(
+        // "0:",
+        // "li t4, 0x43",
+        // "li t5, 0x12440000",
+        // "sw t4, 0(t5)",
+        // "j 0b", // debug: CCCCCCCCCCC
+        // Clear feature disable CSR
+        "csrwi  0x7c1, 0",
+
+        "csrw   mie, zero",
+        "csrw   mstatus, zero",
+        // 2. initialize programming language runtime
+        // clear bss segment
+        "la     t0, sbss",
+        "la     t1, ebss",
+        "1:",
+        "bgeu   t0, t1, 1f",
+        "sd     x0, 0(t0)",
+        "addi   t0, t0, 4",
+        "j      1b",
+        "1:",
+        // 3. prepare stack
+        "la     sp, {stack}",
+        "li     t0, {stack_size}",
+        "add    sp, sp, t0",
+        // "j _debug",
+        "call   {main}",
+        stack      =   sym BT0_STACK,
+        stack_size = const STACK_SIZE,
+        main       =   sym main,
+        options(noreturn)
+    )
+}
+
+fn nop() {
+    unsafe { riscv::asm::nop() }
+}
+
+fn hello() {
+    uart_write('o');
+    uart_write('r');
+    uart_write('e');
+    uart_write('b');
+    uart_write('o');
+    uart_write('o');
+    uart_write('t');
+    uart_write(' ');
+    // uart_write('🦀');
+    uart_write(0xf0 as char);
+    uart_write(0x9f as char);
+    uart_write(0xa6 as char);
+    uart_write(0x80 as char);
+    uart_write('\r');
+    uart_write('\n');
+}
+
+fn main() {
+    clock_init();
+    // for illegal instruction exception
+    crate::init::syscon_func_18(0x00c000c0);
+    rstgen_init();
+
+    // enable core (?)
```

Comment:
This might means to enable another hart. If I'm correct visionfive v1 has two cores, one if faster with virtual memory and the other is only MCU capable. This function might enable the other one to have better preparations

Reply:
Sorry, this comment needs more elaboration. So it's regarding the next line, which is just like the rest here taken from secondBoot, specifically:
https://github.com/starfive-tech/JH7100_secondBoot/blob/e17302063c9a4b74475b18ff24dd149c27257354/boot/bootmain.c#L178

Underneath, it's setting the last bit in `SYSCON_SYSMAIN_CTRL69` (`0x1185_0114`), which we do not have documentation for. 😬